pallets/api/src/messaging/mod.rs

```diff
@@ -608,44 +626,129 @@ impl<T: Config> crate::Read for Pallet<T> {
 #[scale_info(skip_type_params(T))]
 enum Message<T: Config> {
 	Ismp {
 		commitment: H256,
-		callback: Option<Callback>,
-		deposit: BalanceOf<T>,
-	},
-	IsmpTimedOut {
-		commitment: H256,
+		callback: Option<Callback<T::AccountId>>,
 		deposit: BalanceOf<T>,
+		status: MessageStatus,
 	},
 	IsmpResponse {
 		commitment: H256,
 		deposit: BalanceOf<T>,
 		response: BoundedVec<u8, T::MaxResponseLen>,
+		status: MessageStatus,
 	},
 	XcmQuery {
 		query_id: QueryId,
-		callback: Option<Callback>,
+		callback: Option<Callback<T::AccountId>>,
 		deposit: BalanceOf<T>,
+		status: MessageStatus,
 	},
 	XcmResponse {
 		query_id: QueryId,
 		deposit: BalanceOf<T>,
 		response: Response,
+		status: MessageStatus,
 	},
 }
 
+impl<T: Config> Message<T> {
+	/// Try and remove self.
+	pub fn try_remove(&self, origin: &AccountIdOf<T>, id: &MessageId) -> Result<(), DispatchError> {
+		match self {
+			// Ismp messages can only be removed if their status is erroneous.
+			Message::Ismp { status, .. } => match status {
+				MessageStatus::Ok => Err(Error::<T>::RequestPending.into()),
+				MessageStatus::Timeout | MessageStatus::Err(_) => {
+					self.remove(origin, id);
+					Ok(())
+				},
+			},
+			// Ismp responses can always be removed.
+			Message::IsmpResponse { .. } => {
+				self.remove(origin, id);
+				Ok(())
+			},
+			// Xcm queries can only be removed if their status is erroneous.
+			Message::XcmQuery { status, .. } => match status {
+				MessageStatus::Ok => Err(Error::<T>::RequestPending.into()),
+				MessageStatus::Timeout | MessageStatus::Err(_) => {
+					self.remove(origin, id);
+					Ok(())
+				},
+			},
+			// XCM responses can always be removed.
+			Message::XcmResponse { .. } => {
+				self.remove(origin, id);
+				Ok(())
+			},
+		}
+	}
+
+	/// Remove a message from storage.
+	/// Does no check on wether a message should be removed.
+	pub(crate) fn remove(&self, origin: &AccountIdOf<T>, id: &MessageId) {
+		Messages::<T>::remove(&origin, &id);
+		match self {
+			Message::Ismp { commitment, .. } => {
+				IsmpRequests::<T>::remove(commitment);
+			},
+			Message::IsmpResponse { commitment, .. } => {
+				IsmpRequests::<T>::remove(commitment);
+			},
+			Message::XcmQuery { query_id, .. } => {
+				XcmQueries::<T>::remove(query_id);
+			},
+			Message::XcmResponse { query_id, .. } => {
+				XcmQueries::<T>::remove(query_id);
+			},
+		}
+	}
+
+	pub fn release_deposit(&self, origin: &T::AccountId) -> Result<(), DispatchError> {
+		let deposit = match self {
+			Message::Ismp { deposit, .. } => deposit,
+			Message::IsmpResponse { deposit, .. } => deposit,
+			Message::XcmQuery { deposit, .. } => deposit,
+			Message::XcmResponse { deposit, .. } => deposit,
+		};
+		T::Deposit::release(&HoldReason::Messaging.into(), origin, *deposit, Exact)?;
+		Ok(())
+	}
+}
+
+#[derive(Clone, Debug, Encode, Eq, Decode, MaxEncodedLen, PartialEq, TypeInfo)]
+pub enum MessageStatus {
```

Comment:
Nice! This feels better and keeps the `Message` struct a little cleaner. Also makes sense as it's easier to track what the original request was.